helix-core/src/main/java/org/apache/helix/controller/rebalancer/strategy/AutoRebalanceStrategy.java

```diff
@@ -129,9 +129,8 @@ public ZNRecord computePartitionAssignment(final List<String> allNodes, final Li
       logger.info("orphan = " + _orphaned);
     }
 
-    moveNonPreferredReplicasToPreferred();
-
     assignOrphans();
+    moveNonPreferredReplicasToPreferred();
```

Comment:
Because assignOrphans() will also assign some new partitions to non-preferred nodes.